### PR TITLE
[Calyx]Add invoke operation.

### PR DESCRIPTION
**Pull request by linuxlonelyeagle**
This PR only contains `calyx.invoke` and `invoke emitter`.`lower invoke` parts I will submit the other PR.

### CODE REVIEW

**Comment by linuxlonelyeagle** [comment]
@rachitnigam  I resubmitted the PR here, this PR only contains invoke. I think you can review the code now.

File: test/Dialect/Calyx/emit-invoke.mlir
Comment (rachitnigam) [inline test/Dialect/Calyx/emit-invoke.mlir:30]: This CHECK seems incorrect: shouldn't we expect the output to be `r.out` instead of just `out`?

File: test/Dialect/Calyx/errors.mlir
Comment (rachitnigam) [inline test/Dialect/Calyx/errors.mlir:1063]: Can we make this error message more specific and say exactly how many inputs were expected for invocation? In general, an `invoke` is not required to provide assignments for all ports of the component so this is a bit confusing.

File: test/Dialect/Calyx/errors.mlir
Comment (rachitnigam) [inline test/Dialect/Calyx/errors.mlir:1079]: Error message should specify which port it is talking about

File: test/Dialect/Calyx/errors.mlir
Comment (rachitnigam) [inline test/Dialect/Calyx/errors.mlir:1095]: `r` does not have a `go` port so it's not clear what this error message is referring to

File: test/Dialect/Calyx/errors.mlir
Comment (rachitnigam) [inline test/Dialect/Calyx/errors.mlir:1128]: `add` is a combinational component and cannot be invoked. This error message is a bit misleading: it seems to imply that the user could `invoke` `add` if only they provided an `go` and a `done` port.

File: test/Dialect/Calyx/errors.mlir
Comment (rachitnigam) [inline test/Dialect/Calyx/errors.mlir:1144]: What is this error message signaling? That the invoke should provide all ports of the invoked component? In general, that is not true; only the `go` and `done` ports of an invoke are needed.

**Review by rachitnigam — COMMENTED** [review]
Thanks for splitting up the PRs @linuxlonelyeagle! I've left a bunch of comments but at a high-level, there are some clarifications needed:
* Combinational components like `add` cannot be invoked so the error message should signal that.
* The only required ports for an invoke are the `go` and `done` ports (maybe `clk` and `reset` too). Other data ports are considered optional and not always required.
* The error messages should specify which port they're talking about.

Added review requests from @cgyurgyik and @mortbopet to make sure this conforms to the LLVM styleguide

File: test/Dialect/Calyx/emit-invoke.mlir
Comment (linuxlonelyeagle) [reply test/Dialect/Calyx/emit-invoke.mlir:30]: This is correct.`out = out`,the `out` on the `left` represents the `r.out`. The `out` on the `right` is the `out port` of the `main component`.

File: test/Dialect/Calyx/errors.mlir
Comment (rachitnigam) [reply test/Dialect/Calyx/errors.mlir:1063]: Ah, you're right! The `out = out` threw me off.

File: test/Dialect/Calyx/errors.mlir
Comment (linuxlonelyeagle) [reply test/Dialect/Calyx/errors.mlir:1144]: > What is this error message signaling? That the invoke should provide all ports of the invoked component? In general, that is not true; only the `go` and `done` ports of an invoke are needed.

What you can see is that we are invoking r0, but the port used here is r1.in.

**Comment by linuxlonelyeagle** [comment]
* The only required ports for an invoke are the go and done ports (maybe clk and reset too). Other data ports are considered optional and not always required.

I have made changes to what you said.For your second point above, I do this by first checking if a component has either a go or done port (wirte_en and done ports if it's a register), then I check all the invoke port connections to check if the connection involves go or done ports, and then I will check whether the connection involves a go or done port. If there is a go or done port, an error will be reported, because a connection to go and a done port will be generated automatically in pass. In addition, I will also check the destination port and source port of a connection, if neither port is the port of the called component, that indicates that the connection is unrelated to the called component, which will also report an error.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2736]: zip

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2753]: If this is specifically for a hw.module.extern, then why doesn't it take a `HWModuleExternOp` as an argument?
Secondly, the comment to this function doesn't really match with what the function actually does - that is, taking a `StringRef string` and then checking for that string - is it because the `string` is the identifier for `go` or `done`? better naming and docs here :).

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2772]: missing test

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2776]: `one of **the** input or ...`

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2784]: missing braces

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (rachitnigam) [inline lib/Dialect/Calyx/CalyxOps.cpp:2799]: Also, shouldn't this be `calyx.done`?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (linuxlonelyeagle) [reply lib/Dialect/Calyx/CalyxOps.cpp:929]: I'm a little confused about what you mean.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (linuxlonelyeagle) [reply lib/Dialect/Calyx/CalyxOps.cpp:1461]: > It doesn't look like you're using anything from the specific operation type, so this could probably just be `LogicalResult verifyPortDirection(Operation* op, ...` instead of a template.

I use this to verify the connection of the port in the invoke, with this function my first pass parameter is `InvokeOp` and it works fine, I don't think it is necessary to rewrite a function to determine the direction of the port, so I use the original function here.

File: test/Dialect/Calyx/errors.mlir
Comment (cgyurgyik) [inline test/Dialect/Calyx/errors.mlir:1079]: Where is this error message being produced?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (linuxlonelyeagle) [reply lib/Dialect/Calyx/CalyxOps.cpp:929]: Thanks!

**Review by cgyurgyik — COMMENTED** [review]
Nice work! Thanks for splitting this up :-)

File: test/Dialect/Calyx/errors.mlir
Comment (linuxlonelyeagle) [reply test/Dialect/Calyx/errors.mlir:1079]: You can see [here](https://github.com/llvm/circt/blob/72672984a7fa4092d2295a010527d647c2893ae3/lib/Dialect/Calyx/CalyxOps.cpp#L1421).

Comment (cgyurgyik) [reply test/Dialect/Calyx/errors.mlir:1079]: Ah thank you, so this existed before your work.

**Comment by linuxlonelyeagle** [comment]
@rachitnigam  @mortbopet  @cgyurgyik Thank you so much for your suggestions, I have completely revised it in response to the suggestions you gave.:smile:

File: include/circt/Dialect/Calyx/CalyxControl.td
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxControl.td:400]: Let's include both in and out connections in the example, since this has a different syntax than the native Calyx language.

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:838]: Can we save `portName.drop_front(2 + callee.size())` as a local variable?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:2813]:
```suggestion
                           << " is not defined as an input port of '@" << callee << "'.";
```

File: test/Dialect/Calyx/errors.mlir
Comment (cgyurgyik) [inline test/Dialect/Calyx/errors.mlir:1197]: These are two completely different errors: 
1. `is a destination port`, and 
2. `a complex logic`. 

I would strongly suggest splitting these up as such.

**Review by cgyurgyik — APPROVED** [review]
Overall looks good to me. Since this is such a large change, let's get a second pair of eyes to review it. Thank you for the patient review.

File: test/Dialect/Calyx/errors.mlir
Comment (linuxlonelyeagle) [reply test/Dialect/Calyx/errors.mlir]: Thanks for the advice.I've made the changes you suggested.

**Comment by linuxlonelyeagle** [comment]
@mortbopet @rachitnigam I'm really sorry to bother you, but it looks like you may have forgotten that there's code to review here.I have made all the changes based on your suggestions.

**Comment by mortbopet** [comment]
@linuxlonelyeagle never be afraid to do a friendly bump reminder - I am (and probably many others are) swamped by notifications so sometimes something gets lost in the mix.

File: include/circt/Dialect/Calyx/Calyx.td
Comment (mortbopet) [inline include/circt/Dialect/Calyx/Calyx.td:80]: I don't think a comment is necessary since this pattern is fairly common.

However, i do prefer the style used here:
https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/HWStructure.td#L54-L57
(i.e. flipped wrt. how you're doing it now).

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp:1430]: It's not so much that, it's rather than since this function isn't dependent on anything specific from either `InvokeOp` or `AssignOp`, you're solely relying on `Operation*` when you're using this op (calling `op->emitOpError`). In that case, a template function is inappropriate seeing as it will emit two separate functions (once for each template instance) when only one is needed.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2614]: If you really want a comment here, it should be above the function definition, and not directly inside it.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2653]: As mentioned previously, use `llvm::TypeSwitch` here.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2691]: Use `llvm::TypeSwitch` here.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:2727]: This is a case where you'd be expected to add brackets for readability's sake (https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements).

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:852]: `llvm::interleaveComma`?

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:861]: `llvm::interleaveComma`?

File: include/circt/Dialect/Calyx/CalyxControl.td
Comment (linuxlonelyeagle) [inline include/circt/Dialect/Calyx/CalyxControl.td:42]: @mortbopet  Thanks for the advice,I have made the appropriate changes to what you said.To be honest, I don't really understand the style you gave inside the link, I may not understand what that means.But I think it might be used in a different scenes, you can see here.

**Comment by linuxlonelyeagle** [comment]
@mortbopet  @rachitnigam I think you can continue to review the code.:smile:

**Review by mortbopet — APPROVED** [review]
LGTM. Remember to click the "Resolve conversation" button for the review comments that we make, so that us reviewers (and yourself) can keep track of what things have been addressed in your updates.

**Comment by linuxlonelyeagle** [comment]
@rachitnigam Hello sir, the others have accepted the changes, now I just need you to review the code. Earlier I made changes based on your suggestions and after the changes were done, tagged everyone, is this not correct (since I made the changes based on everyone's suggestions)? If there is something I did wrong, please point it out and I will correct it next time.Thanks!

**Comment by rachitnigam** [comment]
@linuxlonelyeagle can you either: (1) open a fix for the issue, or (2) revert the commit while figuring out what the problem is? We can't have tests on `main` failing.

**Comment by rachitnigam** [comment]
@mortbopet any thoughts on why tests would only fail on windows machine? It seems one of the check tests for `--export-calyx` which just dumps out the native representation. I would not have expected them to differ between platforms

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (seldridge) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:838]: I think the Windows failure is stemming from this. The PR is creating a `StringRef` to a `std::string` which is going to go out of scope. An ASAN build or running with valgrind would likely flag this, too.

Comment (linuxlonelyeagle) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:838]: Thank you!Thank you, sir. I'll verify this.

Comment (dtzSiFive) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:838]: ( ASAN report: https://gist.github.com/dtzSiFive/b2851d9be2a81b2aaee5a2851e7ddf27 )

Comment (cgyurgyik) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:838]: Good catch! It is surprising there is no ASAN build before merge.

**Pull request merged**
